src/core/Form/Combobox/Combobox/Combobox.tsx

```diff
@@ -0,0 +1,595 @@
+          selectedItems: controlledItems,
+        } = prevProps;
+        const { selectedKeys, selectedItems } = prevState;
+        const newSelectedKeys = Object.assign({}, selectedKeys);
+        if (!item.disabled) {
+          if (onItemSelect) {
+            onItemSelect(item.uniqueItemId);
+          }
+          if (!controlledItems) {
+            if (item.uniqueItemId in newSelectedKeys) {
+              delete newSelectedKeys[item.uniqueItemId];
+              const newSelectedItems = selectedItems.filter(
+                (selectedItem) =>
+                  selectedItem.uniqueItemId !== item.uniqueItemId,
+              );
+              if (onItemSelectionsChange) {
+                onItemSelectionsChange(newSelectedItems);
+              }
+              return {
+                selectedKeys: newSelectedKeys,
+                selectedItems: newSelectedItems,
+              };
+            }
+            newSelectedKeys[item.uniqueItemId] = false;
+            const newSelectedItems = selectedItems.concat([item]);
+            if (onItemSelectionsChange) {
+              onItemSelectionsChange(newSelectedItems);
+            }
+            return {
+              selectedKeys: newSelectedKeys,
+              selectedItems: newSelectedItems,
+            };
+          }
+        }
+      },
+    );
+  };
+
+  handleRemoveAllSelections = () => {
+    this.setState(
+      (
+        prevState: ComboboxState<T & ComboboxData>,
+        prevProps: ComboboxProps<T & ComboboxData>,
+      ) => {
+        const { selectedItems } = prevState;
+        const { onItemSelectionsChange, onRemoveAll } = prevProps;
+        const disabledItems = [];
+        const newSelectedKeys: ItemKeys = {};
+        if (onRemoveAll) {
+          onRemoveAll();
+        }
+        // eslint-disable-next-line no-restricted-syntax
+        for (const item of selectedItems) {
+          if (item.disabled) {
+            disabledItems.push(item);
+            newSelectedKeys[item.uniqueItemId] = false;
+          }
+        }
+        if (onItemSelectionsChange) {
+          onItemSelectionsChange(disabledItems);
+        }
+        return { selectedKeys: newSelectedKeys, selectedItems: disabledItems };
+      },
+    );
+  };
+
+  render() {
+    const highlightQuery = (text: string, query: string) => {
+      const substrings = text.split(new RegExp(`(${query})`, 'gi'));
+      return substrings.map((substring, i) => {
+        const isMatch = substring.toLowerCase() === query.toLowerCase();
+        if (isMatch) {
+          // eslint-disable-next-line react/no-array-index-key
+          return <mark key={i}>{substring}</mark>;
+        }
+        // eslint-disable-next-line react/no-array-index-key
+        return <span key={i}>{substring}</span>;
+      });
+    };
+
+    const filter = (data: ComboboxData, query: string) =>
+      data.labelText.toLowerCase().includes(query.toLowerCase());
+
+    const handleKeyDown = (event: React.KeyboardEvent) => {
```

Comment:
I propose we create utils.ts for handling keyboard and maybe even state.ts for handling state. This file seems a bit cumbersome to read already.

Reply:
I would like to see able to make it more simple and easier to read, but the handling of keydown and state are so tightly entwined.
I suggest that we could improve this in separate PR to increase readability and maintainability and to not delay this PR with these changes.